### PR TITLE
Feat/create blacklist screen

### PR DESCRIPTION
**Pull request by hellypi**
Add blacklist page so user can blacklist specific cryptocurrency addresses.

### CODE REVIEW

File: lib/blacklistQueries.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+const db = require('./db')
```
Comment (RafaelTaranto) [inline lib/blacklistQueries.js]: There's no reason to rename this file. You can revert the name back to blacklist.js

File: lib/cash-in/cash-in-tx.js [hunk 2]
```diff
@@ -2,7 +2,7 @@ const _ = require('lodash/fp')
 const pgp = require('pg-promise')()
```
Comment (RafaelTaranto) [inline lib/cash-in/cash-in-tx.js]: This whole file was changed only because of automatic formatting that's not really configured and of the blacklist file rename. You can undo all changes here.

File: lib/new-admin/graphql/schema.js [hunk 3]
```diff
@@ -222,18 +247,28 @@ const typeDefs = gql`
...
+      cassette2: Int
+      newName: String
+    ): Machine
     machineSupportLogs(deviceId: ID!): SupportLogsResponse
     serverSupportLogs: SupportLogsResponse
     setCustomer(customerId: ID!, customerInput: CustomerInput): Customer
     saveConfig(config: JSONObject): JSONObject
     createPairingTotem(name: String!): String
     saveAccounts(accounts: JSONObject): JSONObject
+    deleteBlacklistRow(cryptoCode: String, address: String): [String]
```
Comment (RafaelTaranto) [inline lib/new-admin/graphql/schema.js]: Don't we want cryptoCode and address to be required here as well? Is there are reason why this returns an array?

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            saveConfig={saveConfig}
+            configData={configData}
+          />
+        </div>
+      </Grid>
+      <BlackListModal
+        showModal={showModal}
+        toggleModal={toggleModal}
+        selectedCoin={clickedItem}
+        addToBlacklist={addToBlacklist}></BlackListModal>
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: This can be self closing

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  }
+`
+
+const Blacklist = () => {
+  const { data: blacklistResponse } = useQuery(GET_BLACKLIST)
+  const { data: configData } = useQuery(GET_INFO)
+  const blacklistData = R.path(['blacklist'])(blacklistResponse) ?? []
+  const availableCurrencies =
+    R.path(['cryptoCurrencies'], blacklistResponse) ?? []
+
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: Hooks have a bunch of rules for them to work properly, because of this I tend to prefer to get group them on the top of the function. 

Try to move all hooks to the top and the logic related to them afterwards.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+  const classes = useStyles()
+
+  const [showModal, setShowModal] = useState(false)
+  const [clickedItem, setClickedItem] = useState({
+    code: 'BTC',
+    display: 'Bitcoin'
+  })
+
+  const groupByCode = R.groupBy(obj => obj.cryptoCode)
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: Another thing that helps with readability is moving pure functions out of the component, this one can be outside the Blacklist function definition.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    refetchQueries: () => ['getBlacklistData']
+  })
+
+  const [saveConfig] = useMutation(SAVE_CONFIG, {
+    refetchQueries: () => ['getData']
+  })
+
+  const isSelected = it => {
+    return clickedItem.code === it.code
+  }
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: you can remove the brackets and the return to shorten this up

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+
+  const toggleModal = () => {
+    setShowModal(!showModal)
+  }
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: you can remove brackets here as well

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+          displayName={it => it.display}
+          onClick={onClickSidebarItem}
+        />
+        <div className={classes.content}>
+          <BlacklistTable
+            data={formattedData}
+            selectedCoin={clickedItem}
+            handleDeleteEntry={handleDeleteEntry}
+            saveConfig={saveConfig}
+            configData={configData}
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: We can probably simplify this into moving the logic to the outside of the BlacklistTable component. The idea is to make the component as dumb as possible with relation to the context. 

Instead of passing down the configData as a whole you can just pass down rejectAddressReuse and have Blacklist.js figure out the namespacing logic, the same is true for saving.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 10]
```diff
@@ -0,0 +1,42 @@
+export default {
+  grid: {
+    flex: 1,
+    height: '100%'
+  },
+  content: {
+    display: 'flex',
+    flexDirection: 'column',
+    flex: 1,
+    marginLeft: 48
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: We're not really using this everywhere but it's good practice to use our `spacer` variable. You can import it from 'src/styling/variables'. 

Then you can make it spacer * 6 instead of 48.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  bottomRight: {
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: This can do with a different name. Like `footer`.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+    marginLeft: 'auto',
+    marginTop: 'auto',
+    marginBottom: '24px'
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: this can be 
`margin: [['auto', 0, spacer * 3, 'auto']]`

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  modalTitle: {
+    lineHeight: '120%',
+    color: '#1b2559',
+    fontSize: 14,
+    fontFamily: 'Mont',
+    fontWeight: 900
+  },
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: Isn't this already defined in typography? We don't want to manually set font styles nor use unnamed colors. The colors are all exported with names from the styling/variables file.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  subtitle: {
+    display: 'flex',
+    justifyContent: 'space-between',
+    alignItems: 'center',
+    flexDirection: 'row'
+  },
+  white: {
+    color: 'white'
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: even though it shouldn't be a big deal, we're exporting white as a named variable as well. The reasoning for naming everything instead of using a const is that if we decide to slightly change our base white color the change will be applied to all places if we're using the variable name.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  deleteButton: {
+    paddingLeft: '13px'
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: You can shorten this to

`paddingLeft: 13` you can use numbers instead of strings with `px` at the end.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  addressRow: {
+    marginLeft: '8px'
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: same thing as above

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+  },
+  table: {
+    height: '900px'
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.styles.js]: Any reason for having this specified height on table? We probably want to to follow the size of the page.

File: new-lamassu-admin/src/routing/routes.js [hunk 18]
```diff
@@ -10,6 +10,8 @@ import {
 
 import { AppContext } from 'src/App'
 import AuthRegister from 'src/pages/AuthRegister'
+// import should appear before cashout (why?)
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/routing/routes.js]: You can remove this comment. I think I already explained this in a call, but imports are organized by scope and alphabetically.

File: new-lamassu-admin/src/routing/routes.js [hunk 19]
```diff
@@ -147,6 +149,13 @@ const tree = [
         route: '/compliance/customers',
         component: Customers
       },
+      {
+        key: 'blacklist',
+        label: 'Blacklist',
+        route: '/compliance/blacklist',
+        component: Blacklist
+      },
+      // what is this last one for
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/routing/routes.js]: What is this comment?

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: new-lamassu-admin/src/routing/routes.js (hunk 19 above)
Comment (hellypi) [reply new-lamassu-admin/src/routing/routes.js]: That was a question I wanted to ask you but ended up forgetting. I started taking notes on the side instead of commenting, that one escaped my eye...

**Comment by hellypi** [comment]
[Create the blacklist screen](https://trello.com/c/OjxjUUH7/165-create-the-blacklist-screen)

**Comment by csrapr** [comment]
[Create the blacklist screen](https://trello.com/c/OjxjUUH7/165-create-the-blacklist-screen)

File: lib/blacklist.js [hunk 20]
```diff
@@ -1,22 +1,52 @@
...
+    .any(
+      'insert into blacklist(crypto_code, address, created_by_operator) values($1, $2, $3);',
+      [cryptoCode, address, true]
+    )
+    .then(() => {
+      return { cryptoCode, address }
+    })
+}
+
+function blocked(address, cryptoCode) {
```
Comment (RafaelTaranto) [inline lib/blacklist.js:33]: Hmm, this doesn't pass standard. Functions should have a space between the name and parenthesis on function definition.

File: lib/blacklist.js [hunk 21]
```diff
@@ -1,22 +1,52 @@
...
   const sql = `select * from blacklist where address = $1 and crypto_code = $2`
-  return db.any(sql, [
-    address,
-    cryptoCode
-  ])
+  return db.any(sql, [address, cryptoCode])
 }
 
-function addToUsedAddresses (address, cryptoCode) {
+function addToUsedAddresses(address, cryptoCode) {
```
Comment (RafaelTaranto) [inline lib/blacklist.js:38]: Same thing here

File: lib/cash-in/cash-in-tx.js [hunk 22]
```diff
@@ -176,4 +176,4 @@ function cancel (txId) {
       if (res.rowCount !== 1) throw new Error('No such tx-id')
     })
     .then(() => logActionById('operatorCompleted', {}, txId))
-}
+}
```
Comment (RafaelTaranto) [inline lib/cash-in/cash-in-tx.js:179]: Undo the change of this file.

File: new-lamassu-admin/src/utils/coins/bch.js [hunk 23]
```diff
@@ -0,0 +1,15 @@
+const cashaddr = require('cashaddrjs')
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/utils/coins/bch.js]: I don't see this added on package.json, this won't really work.

It might be too annoying to fix or move to the server side for all the addresses, remove all validation code.

We can circle back on this if the need arises.

File: new-lamassu-admin/src/pages/Triggers/Triggers.js [hunk 24]
```diff
@@ -38,7 +38,7 @@ const Triggers = () => {
 
   const { data, loading } = useQuery(GET_INFO)
   const triggers = fromServer(data?.config?.triggers ?? [])
-
+  console.log(triggers)
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Triggers/Triggers.js]: undo changes to this file

File: new-lamassu-admin/src/pages/Blacklist/BlacklistTable.js [hunk 25]
```diff
@@ -0,0 +1,62 @@
...
+        </IconButton>
+      )
+    }
+  ]
+  const dataToShow = selectedCoin
+    ? data[selectedCoin.code]
+    : data[R.keys(data)[0]]
+
+  return (
+    <>
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistTable.js]: If you're returning just a single node you don't really need the `<> </>`

File: new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js [hunk 26]
```diff
@@ -0,0 +1,99 @@
...
+            autoFocus
+            id="address-to-block-input"
+            type="text"
+            size="sm"
+            fullWidth
+            InputLabelProps={{ shrink: true }}
+            placeholder={`ex: ${placeholderAddress[selectedCoin.code]}`} // "ex: 0x309abbd2f85ead2fcbb5323e963550c88c8a569aca4e088e9020a03fd04bf4bd"
+            onChange={handleChange}
+            value={addressField}
+          />
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js]: Change this to use formik, you can create use simple validation since we're removing the coin specific parts.

File: new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js [hunk 27]
```diff
@@ -0,0 +1,99 @@
...
+  const placeholderAddress = {
+    BTC: '1ADwinnimZKGgQ3dpyfoUZvJh4p1UWSSpD',
+    ETH: '0x71C7656EC7ab88b098defB751B7401B5f6d8976F',
+    LTC: 'LPKvbjwV1Kaksktzkr7TMK3FQtQEEe6Wqa',
+    DASH: 'XqQ7gU8eM76rEfey726cJpT2RGKyJyBrcn',
+    ZEC: 't1KGyyv24eL354C9gjveBGEe8Xz9UoPKvHR',
+    BCH: 'qrd6za97wm03lfyg82w0c9vqgc727rhemg5yd9k3dm'
+  }
+
+  coinValidator.BTC('HELLO?')
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js]: Remove this

File: new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js [hunk 28]
```diff
@@ -0,0 +1,99 @@
...
+
+  const [addressField, setAddressField] = useState('')
+  const [invalidAddress, setInvalidAddress] = useState(false)
+
+  const handleChange = event => {
+    if (event.target.value === '') {
+      setInvalidAddress(false)
+    }
+    setAddressField(event.target.value)
+  }
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js]: Formik should handle both validation code and field state

File: new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+      setInvalidAddress(true)
+      return
+    }
+    if (coinValidator[selectedCoin.code](addressField.trim())) {
+      addToBlacklist(selectedCoin.code, addressField.trim())
+      handleClose()
+    } else {
+      setInvalidAddress(true)
+    }
+  }
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js]: We don't really wanna close the modal before getting the response from the server. See how other wizards handle this.

File: new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js [hunk 30]
```diff
@@ -0,0 +1,99 @@
...
+
+  const handleClose = () => {
+    setAddressField('')
+    setInvalidAddress(false)
+    toggleModal()
+  }
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js]: State should be handled by formik and handleClose should call a external close handler.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 31]
```diff
@@ -0,0 +1,183 @@
...
+`
+
+const Blacklist = () => {
+  const { data: blacklistResponse } = useQuery(GET_BLACKLIST)
+  const { data: configData } = useQuery(GET_INFO)
+  const [showModal, setShowModal] = useState(false)
+  const [clickedItem, setClickedItem] = useState({
+    code: 'BTC',
+    display: 'Bitcoin'
+  })
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js:75]: Easier to just set the code as state, no reason to have a object here.

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 32]
```diff
@@ -0,0 +1,183 @@
...
+
+  return (
+    <>
+      <TitleSection title="Blacklisted addresses">
+        <Link onClick={toggleModal}>Blacklist new addresses</Link>
+      </TitleSection>
+      <Grid container className={classes.grid}>
+        <Sidebar
+          data={availableCurrencies}
+          isSelected={isSelected}
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: You can probably do R.propEq here instead of defining a isSelected function

File: new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js [hunk 33]
```diff
@@ -0,0 +1,99 @@
...
+    DASH: 'XqQ7gU8eM76rEfey726cJpT2RGKyJyBrcn',
+    ZEC: 't1KGyyv24eL354C9gjveBGEe8Xz9UoPKvHR',
+    BCH: 'qrd6za97wm03lfyg82w0c9vqgc727rhemg5yd9k3dm'
+  }
+
+  coinValidator.BTC('HELLO?')
+
+  return (
+    <>
+      {showModal && (
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/BlacklistModal.js]: Better to have showModal handled on the outside component

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js [hunk 34]
```diff
@@ -0,0 +1,183 @@
...
+          <BlacklistTable
+            data={formattedData}
+            selectedCoin={clickedItem}
+            handleDeleteEntry={handleDeleteEntry}
+          />
+        </div>
+      </Grid>
+      <BlackListModal
+        showModal={showModal}
+        toggleModal={toggleModal}
```
Comment (RafaelTaranto) [inline new-lamassu-admin/src/pages/Blacklist/Blacklist.js]: Name this onClose or similar, we don't need to ever toggle the modal inside the modal, pass down the `() => setShowModal(false)` to this

**Review by RafaelTaranto — CHANGES_REQUESTED** [review]

File: new-lamassu-admin/src/utils/coins/bch.js (hunk 23 above)
Comment (hellypi) [reply new-lamassu-admin/src/utils/coins/bch.js]: all validation code for all coins or just for BCH?

File: new-lamassu-admin/src/pages/Blacklist/Blacklist.js (hunk 31 above)
Comment (hellypi) [reply new-lamassu-admin/src/pages/Blacklist/Blacklist.js:75]: as we talked in private message, we do need this after all 👍 :)
